src/operators.jl

```diff
@@ -244,65 +267,161 @@ function flatten(input::Signal; typ=Any, name=auto_name!("flatten", input))
+                        #so we must ignore nodes whose ancestors without this node
+                        #contain deadroot
+                        node = action.recipient.value
+                        deadroot in roots_without(node, output) && continue
+                        remove_actions!(node, deadroot)
+                    end
+                end
+                current_node = value(input)
+                send_value!(output, value(current_node))
+                wire_flatten(current_node, subtree_actions)
+            end
+        end
+    end
+end
 
-        add_action!(callback, current_node, output)
-
-        add_action!(input, output) do output, timestep
-
-            # Move around action from previous node to current one
-            remove_action!(callback, current_node, output)
-            current_node = value(input)
-            add_action!(callback, current_node, output)
+queue_subtree_actions(queue, basenode) = begin
+    baseidx = find(action->action.recipient.value == basenode, queue) |> first
+    subtree_nodes = Any[basenode]
+    subtree_actions = Action[queue[baseidx]]
+    for actionidx in baseidx:length(queue)
+        action = queue[actionidx]
+        node = action.recipient.value
+        if any(map(node.parents) do node; node in subtree_nodes end)
+            #node has parents that are in the sub-tree
+            push!(subtree_nodes, node)
+            push!(subtree_actions, action)
+        end
+    end
+    subtree_actions
+end
 
-            send_value!(output, value(current_node), timestep)
+"""
+find roots in paths not through ignorenode
+"""
+roots_without(startnode, ignorenode; roots = Dict{Signal, Bool}()) = begin
+    goodparents = filter(startnode.parents) do parent; parent != ignorenode end
+    for parent in goodparents
+        if isempty(parent.parents)
+            roots[parent] = true
+        else
+            roots_without(parent, ignorenode; roots=roots)
         end
     end
+    keys(roots)
 end
 
 const _bindings = Dict()
 
 """
-    bind!(a,b,twoway=true)
+    `bind!(dest, src, twoway=true)`
 
-for every update to `a` also update `b` with the same value and vice-versa.
-To only bind updates from b to a, pass in a third argument as `false`
-"""
-function bind!(a::Signal, b::Signal, twoway=true)
+for every update to `src` also update `dest` with the same value and, if
+`twoway` is true, vice-versa.
 
-    let current_timestep = 0
-        action = add_action!(b, a) do a, timestep
-            if current_timestep != timestep
-                current_timestep = timestep
-                send_value!(a, value(b), timestep)
-            end
+Note that signals dependent on `dest` added before calling `bind!` (e.g. those
+created by calling `map(f, dest)`, `merge(dest, s1)`, `filter(f, dest)`, etc.
+before `bind!(src, dest)`) may be updated more than once when `src` updates. To
+ensure signals update just once per `push!`, call `bind!` before adding signals
+dependent on `dest`.
```

Comment:
I don't think so, this was a change late in the day, I should add it to the downsides on the PR notes once I've checked it. 

Note this is only a problem if someone does the thing that the little note says not to do, and has a downstream op that has some side-effect that they don't want running more than once per-push, or perhaps some expensive op that would run twice and thus impact performance.